Fix horovod build failure when mxnet is built from source

apeforest opened the pull request:
## Description ##
Fix https://github.com/apache/incubator-mxnet/issues/15211

The problem was due to the missing mkldnn_version.h file under include/mkldnn.

The mkldnn_version.h is generated at compile time by MKLDNN library. It is linked to 3rdparty/mkldnn/build/include when mxnet is built using make, the default build system currently supported in MXNet.

An earlier PR https://github.com/apache/incubator-mxnet/pull/14899 fixed the issue when mxnet is installed by pip. This PR is to fix the build error when mxnet is built from source.

Thanks @TaoLv for rootcausing the problem and his suggestions.

apeforest commented:
@PatricZhao @yuxihu @TaoLv Please help to review.

TaoLv commented:
Seems cmake is not happy with the change :(

apeforest commented:
> Seems cmake is not happy with the change :(

Unfortunately this is a hotfix that can help user to build horovod with mxnet from source. We can ask users to build using `make`, which I think is still the de facto build system of mxnet.

yuxihu commented on include/mkldnn/mkldnn.h:
```diff
@@ -0,0 +1 @@
+../../3rdparty/mkldnn/include/mkldnn.h
```
why do we need to link this file? We only miss mkldnn_version.h, right?

apeforest replied:
Earlier it was linked to the entire 3rdparty/mkldnn/include folder. However it does not contain mkldnn_version.h there. In this change, I only link to required files instead of linking to the entire folder.

yuxihu replied:
Did you verify if it works? I remember mkldnn.h may depend on other header files in the original folder.

apeforest replied:
Yes, I could verify it works fine in my local branch.

yuxihu replied:
Great. Please retrigger the CI.

TaoLv replied:
I see the including chain is: ndarray.h -> mkldnn.hpp -> mkldnn.h -> mkldnn_types.h and mkldnn_version.h.

roywei commented:
I don't think re-trigger CI is going to pass, is there any way to fix CMake errors in CI?

apeforest commented:
@TaoLv This is not working because there are two issues in CI:
(1) ASAN is using Cmake and it will break
(2) CI lint complains about mxnet_version.h. Since it's a generated file at compile time, I cannot fix it.

After some more thoughts, can we just modify the make script in MKLDNN module to automatically copy mkldnn_version.h to the include folder after compilation is complete? This way it solves problem for both cmake and make.

This issue is not limited to horovod. Any downstream project that needs to build mxnet from source will fail due to this. 

@PatricZhao Your advice is highly appreciated. Thanks!

marcoabreu commented:
My two cents are that we must not break make as well as cmake. Both are supported and should be kept.

TaoLv commented:
@apeforest I think we need add an install target to the Makefile and copy libraries and including files to the install destination. Downstream projects like horovod should find libraries and headers there, not in the source code of mxnet or mkldnn. @szha @marcoabreu 

```
install:
       mkdir -p $(INSTALL_PREFIX)/lib/mxnet
       mkdir -p $(INSTALL_PREFIX)/include/mxnet
       cp -rf lib/* $(INSTALL_PREFIX)/lib/mxnet/
       cp -Lrf include/* $(INSTALL_PREFIX)/include/mxnet/

       if [ -e "lib/libmkldnn.so.0" ]; then \
               mkdir -p $(INSTALL_PREFIX)/include/mxnet/mkldnn; \
               cp -rf 3rdparty/mkldnn/build/install/include/* $(INSTALL_PREFIX)/include/mxnet/mkldnn/; \
       fi
```

apeforest commented:
@TaoLv There is an issue with this approach:
When Horovod builds with mxnet, it actually uses `mxnet.libinfo.find_include_path()` to look for the header file location. With your suggested change, how can mxnet know the include path at runtime?

TaoLv commented:
My two cents:
1. If you really want to copy files in source code folder, you can do that in the `mkldnn.mk` file. It will help to leave MXNet Makefile clean.
2. The change will generate new folder and file under include/mxnet folder which will be shown as untracked files when call `git status` and the new folder and file cannot be removed by `make clean`.

apeforest commented:
@TaoLv  Thanks for your suggestion.
1. I updated mkldnn.mk to leave Makefile clean
2. I updated .gitignore file so that you would not see untracked files in include/mkldnn
3. It can be cleaned by make clean.

Please help to review again.

vandanavk commented:
@mxnet-label-bot add [pr-awaiting-review, Distributed]

roywei commented:
Thanks for the fix. I have verified this works
mxnet:
```
sudo make -j$(nproc) USE_CUDA=1 USE_CUDNN=1 USE_CUDA_PATH=/usr/local/cuda/ USE_NCCL=1
```
horovod:
```
 HOROVOD_GPU_ALLREDUCE=NCCL  pip install horovod --user --no-cache-dir -U
```
result:
```
Collecting horovod
  Downloading https://files.pythonhosted.org/packages/42/f8/0a2fedf45122d8a1b2dbd573e737ccb32cd0776aa4c4b157d3f18b9ff0ca/horovod-0.16.4.tar.gz (2.6MB)
    100% |████████████████████████████████| 2.6MB 87.4MB/s 
Requirement already satisfied, skipping upgrade: cloudpickle in /usr/local/lib/python2.7/dist-packages (from horovod) (1.0.0)
Requirement already satisfied, skipping upgrade: psutil in /home/ubuntu/.local/lib/python2.7/site-packages (from horovod) (5.6.3)
Requirement already satisfied, skipping upgrade: six in /usr/local/lib/python2.7/dist-packages (from horovod) (1.10.0)
Requirement already satisfied, skipping upgrade: cffi>=1.4.0 in /home/ubuntu/.local/lib/python2.7/site-packages (from horovod) (1.12.3)
Requirement already satisfied, skipping upgrade: pycparser in /home/ubuntu/.local/lib/python2.7/site-packages (from cffi>=1.4.0->horovod) (2.19)
Installing collected packages: horovod
  Running setup.py install for horovod ... done
Successfully installed horovod-0.16.4
```

`git status`
```
git status
On branch bugfix/horovod_build
Your branch is up-to-date with 'origin/bugfix/horovod_build'.
nothing to commit, working directory clean```

`sudo make clean` also removes it

TaoLv commented on tools/license_header.py:
```diff
@@ -68,6 +68,9 @@
                '3rdparty',
 
+               # 3rdparty headerfiles under different licenses
+               'include/mkldnn',
```
Can you explain? MKL-DNN is also under Apache-2.0.

szha replied:
I imagine mkldnn would be licensed to intel instead of ASF even though the license is apache 2.0

apeforest replied:
license check would fail if this is not whitelisted.

szha commented on include/mkldnn:
```diff
@@ -1 +0,0 @@
-../3rdparty/mkldnn/include
```
why not just change this to the install path?

apeforest replied:
Because we do not need to export all the header files under the install path.

szha replied:
I think it's simpler that way. Otherwise next time if there's any change mkldnn header we need to go through this all over again.

roywei commented:
I have created feature request to better address this #15255

Can we merge this PR if it all looks good?
